### PR TITLE
use Ruff for linting

### LINKED ISSUE #14817: Consider ruff for linting and replace Flake8 with it
### Is your feature request related to a problem? Please describe.
Flake8 is a very popular linter, but it is extrelmely slow and it checks every time all files, even if they have not been changed.
It also lacks of relevant rules and the development activity is slowing down.

### Describe the solution you'd like
Ruff is a very promising alternative, combining Pylint, Flake8 and other linters in one very fast environment. It can be configured so it checks only files that have been changed when sending a pull request.
https://github.com/charliermarsh/ruff
Its development team is very active and already many projects are using it especially because it provides a good mix of many linters out there and still it is very fast.
It has also a flake8 Gettext plugin included, so translatable strings could also be linted.

### Describe alternatives you've considered
Leave as it is.

### Additional context
n/a

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
Fixes #14817 

### Summary of the issue:
Our linting system has the following issues:
- we are stuck on an old version of flake8, as the flake8tabs module is no longer being updated
- we can't perform automatic lint fixes
- Ruff is becoming more maintained than flake8

### Description of user facing changes
- The entire repository is linted with Ruff, rather than diffs
- Developers can now automatically fix certain linting issues with `runlint.bat`
- If using pre-commit hooks, linting is automatically applied: https://pre-commit.com/#usage
- CodeRabbit now scans code using our Ruff config

### Description of development approach

- integrate Ruff, migrate flake8 config
- Update AppVeyor scripts to use ruff
- add pre commit hook support for ruff
- update linting docs
- [x] TODO: migrate wiki page after merge: https://github.com/nvaccess/nvda/wiki/LintWithFlake8
- [x] TODO: lint the codebase after merge: https://github.com/nvaccess/nvda/pull/16766
- [x] TODO: update VS code repo to support ruff : https://github.com/nvaccess/vscode-nvda/commit/1a6191e1f5d0d241ac9f406dede9be69a55b26df

To lint the whole code base:
1. `ruff check --fix` to apply automatic fixes
1. `ruff check --add-noqa` to add no-qa comments to all lint failures, so that issues that can't be automatically fixed don't trigger future failures
1. add these commits to `.git-ignore-blame-revs`

See https://github.com/nvaccess/nvda/pull/16752/ for an example

### Testing strategy:

- [x] Test successful lint with Ruff PR: https://github.com/nvaccess/nvda/pull/16752
- [x] Test pre-commit hooks
- [x] Test AppVeyor build with lint failure: https://ci.appveyor.com/project/NVAccess/nvda/builds/50101991/tests

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Introduced Ruff linter for Python code style enforcement.
	- Added pre-commit hooks for automated linting and formatting.

- **Bug Fixes**
	- Updated NVDA build system for compatibility with Visual Studio 2022 version 17.10 and above.
	- Improved readability in Log Viewer and NVDA Python Console by using a fixed-width font.

- **Documentation**
	- Updated documentation to reflect the switch from Flake8 to Ruff for linting.
	- Added new documentation on integrating Ruff with development tools.

- **Chores**
	- Replaced Flake8 with Ruff in configuration and script files.

### CODE REVIEW

File: appveyor/scripts/tests/lintCheck.ps1 [hunk 1]
```diff
@@ -16,17 +15,14 @@ if ($env:APPVEYOR_PULL_REQUEST_NUMBER -or $env:APPVEYOR_REPO_BRANCH.StartsWith("
 		git fetch -q origin master:master
 		$msgBaseLabel = "Branch"
 	}
-	.\runlint.bat FETCH_HEAD "$flake8Output" 
+	.\runlint.bat "$lintOutput"
```
Comment (seanbudd) [inline appveyor/scripts/tests/lintCheck.ps1:18]: can the above checks be removed?

File: runlint.bat [hunk 2]
```diff
@@ -1,15 +1,11 @@
 @echo off
-rem runlint <base commit> [<output file>]
-rem Lints any changes after base commit up to and including current HEAD, plus any uncommitted changes.
+rem runlint
```
Comment (seanbudd) [inline runlint.bat:2]:
```suggestion
rem runlint [<output file>]
```

File: appveyor/scripts/tests/lintCheck.ps1 (hunk 1 above)
Comment (seanbudd) [reply appveyor/scripts/tests/lintCheck.ps1:18]: no, it should remain

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+repos:
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  # Ruff version.
+  rev: v0.4.10
```
Comment (LeonarddeR) [inline .pre-commit-config.yaml:4]: Could we possibly use master here?
```suggestion
  rev: master
```

File: pyproject.toml
Comment (LeonarddeR) [inline pyproject.toml:1]: For reference, here is the `ruff.toml` I've been using internally. Feel free to investigate and add where appropriate:

```toml
builtins = [
	"_",  # translation lookup
	"ngettext",  # translation lookup
	"pgettext",  # translation lookup
	"npgettext",  # translation lookup
]
logger-objects = ["logHandler.log"]
exclude = [  # don't bother looking in the following subdirectories / files.
	".git",
	"__pycache__",
	".tox",
	"build",
	"output",
	"include/*",
	"miscDeps",
	"source/louis",
	"source/comInterfaces/*",
]
ignore = [
	"E274",  # Tab before keyword
	"W191",  # indentation contains tabs
]
line-length = 110
select = [
	"A",  # flake8-builtins
	"ARG",  # flake8-unused-arguments 
	"B",  # flake8-bugbear
	"C90", # mccabe
	"E",  # pycodestyle Error
	"F",  # Pyflakes
	"FIX",  # flake8-fixme
	"I",  # isort
	"INT",  # flake8-gettext 
	"UP",  # pyupgrade
	"W",  # pycodestyle Warning
]
show-source = true
src = ["source"]
target-version = "py311"

[mccabe]
max-complexity = 15

[format]
line-ending = "lf"
indent-style = "tab"

[flake8-gettext]
extend-function-names = ["npgettetxt"]  # Not included by default

```

Comment (hwf1324) [reply pyproject.toml:1]: This is the Ruff configuration I used a long time ago, just for reference.

```toml
[tool.ruff]
# Enable the pycodestyle (`E`) and Pyflakes (`F`) rules by default.
# Unlike Flake8, Ruff doesn't enable pycodestyle warnings (`W`) or
# McCabe complexity (`C901`) by default.
select = ["C9", "E", "F", "W"]
ignore = ["W191"]

# Allow fix for all enabled rules (when `--fix`) is provided.
fixable = ["ALL"]
unfixable = []

# Exclude a variety of commonly ignored directories.
exclude = [
	".bzr",
	".direnv",
	".eggs",
	".git",
	".git-rewrite",
	".hg",
	".mypy_cache",
	".nox",
	".pants.d",
	".pytype",
	".ruff_cache",
	".svn",
	".tox",
	".venv",
	"__pycache__",
	"__pypackages__",
	"_build",
	"buck-out",
	"build",
	"dist",
	"include/*",
	"miscDeps",
	"node_modules",
	"output",
	"source/louis",
	"source/comInterfaces/*", # #10924: generated by third-party dependencies
	"venv",
]
per-file-ignores = {}

# Same as Black.
line-length = 110

# Allow unused variables when underscore-prefixed.
dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"

# Assume Python 3.11
target-version = "py311"

# inform ruff about functions we consider built-in.
builtins = ["_", "ngettext", "pgettext", "npgettext"] # translation lookup

show-source = true

src = ["source"]

[tool.ruff.mccabe]
max-complexity = 15
```

File: .pre-commit-config.yaml (hunk 3 above)
Comment (seanbudd) [reply .pre-commit-config.yaml:4]: I think we should match requirements.txt and keep it pinned

**Review by gerald-hartig — APPROVED** [review]

**Pull request merged**